File: services/multi-agent-orchestrator/src/telemetry.ts
import {
  SpanKind,
  SpanStatusCode,
  context,
  trace,
  type Context,
  type TextMapGetter,
} from "@opentelemetry/api";
import { W3CTraceContextPropagator } from "@opentelemetry/core";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { NodeSDK } from "@opentelemetry/sdk-node";

export const GEN_AI_SEMCONV_COMMIT = "94f432d7126f5884d30a2cdde6f4e89908ebb6fd";
const propagator = new W3CTraceContextPropagator();
const getter: TextMapGetter<Record<string, string | string[] | undefined>> = {
  keys(carrier) { return Object.keys(carrier); },
  get(carrier, key) { return carrier[key.toLowerCase()]; },
};
let sdk: NodeSDK | undefined;

export async function startTelemetry(environment: NodeJS.ProcessEnv = process.env): Promise<void> {
  if (environment.OTEL_SDK_DISABLED === "true"
      || environment.PLATFORM_OBSERVABILITY_OTLP_ENABLED !== "true") return;
  if (sdk) return;
  const endpoint = environment.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT
    ?? environment.PLATFORM_OBSERVABILITY_OTLP_ENDPOINT;
  if (!endpoint) return;
  try {
    const candidate = new NodeSDK({
      traceExporter: new OTLPTraceExporter({ url: endpoint }),
      textMapPropagator: propagator,
      serviceName: "multi-agent-orchestrator",
    });
    candidate.start();
    sdk = candidate;
  } catch {
    sdk = undefined;
  }
}

export async function shutdownTelemetry(): Promise<void> {
  const current = sdk;
  sdk = undefined;
  if (current) await current.shutdown();
}

export async function withOrchestrationSpan<T>(
  headers: Record<string, string | string[] | undefined>,
  operation: () => Promise<T>,
): Promise<T> {
  const parent = extractContext(headers);
  const tracer = trace.getTracer("spaceagent.multi-agent", "0.1.0");
  return context.with(parent, () => tracer.startActiveSpan(
    "invoke_workflow spaceagent.multi_agent",
    {
      kind: SpanKind.INTERNAL,
      attributes: {
        "gen_ai.operation.name": "invoke_workflow",
        "gen_ai.workflow.name": "spaceagent.multi_agent",
        "spaceagent.contract.version": "multi-agent/v1",
        "spaceagent.gen_ai.semconv.commit": GEN_AI_SEMCONV_COMMIT,
      },
    },
    async (span) => {
      try {
        const result = await operation();
        span.setStatus({ code: SpanStatusCode.OK });
        return result;
      } catch (cause) {
        span.setAttribute("error.type", safeErrorType(cause));
        span.setStatus({ code: SpanStatusCode.ERROR });
        throw cause;
      } finally {
        span.end();
      }
    },
  ));
}

export function extractedTraceId(
  headers: Record<string, string | string[] | undefined>,
): string | undefined {
  return trace.getSpanContext(extractContext(headers))?.traceId;
}

function extractContext(headers: Record<string, string | string[] | undefined>): Context {
  return propagator.extract(context.active(), headers, getter);
}

function safeErrorType(cause: unknown): string {
  if (!(cause instanceof Error)) return "_OTHER";
  const value = cause.name.trim();
  return /^[A-Za-z0-9_.-]{1,120}$/.test(value) ? value : "_OTHER";
}

File: workers/sandbox-worker/sandbox_worker/telemetry.py
"""Official OpenTelemetry boundary with a deliberately redacted GenAI subset."""

from __future__ import annotations

import os
from contextlib import contextmanager
from typing import Iterator, Mapping

from opentelemetry import trace
from opentelemetry.context import Context
from opentelemetry.propagate import extract
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.trace import Span, SpanKind, Status, StatusCode

GEN_AI_SEMCONV_COMMIT = "94f432d7126f5884d30a2cdde6f4e89908ebb6fd"


def configure_telemetry(environment: Mapping[str, str] | None = None) -> None:
    values = environment if environment is not None else os.environ
    if values.get("OTEL_SDK_DISABLED") == "true":
        return
    if values.get("PLATFORM_OBSERVABILITY_OTLP_ENABLED") != "true":
        return
    endpoint = values.get("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT") or values.get(
        "PLATFORM_OBSERVABILITY_OTLP_ENDPOINT"
    )
    if not endpoint:
        return
    try:
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter

        provider = TracerProvider(
            resource=Resource.create({"service.name": "sandbox-worker"})
        )
        provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=endpoint)))
        trace.set_tracer_provider(provider)
    except Exception:  # noqa: BLE001 - telemetry must never affect execution availability
        return


def extracted_trace_id(headers: Mapping[str, str]) -> str | None:
    span = trace.get_current_span(extract(_lower(headers)))
    span_context = span.get_span_context()
    return format(span_context.trace_id, "032x") if span_context.is_valid else None


@contextmanager
def execute_tool_span(headers: Mapping[str, str], tool_name: str) -> Iterator[Span]:
    tool = _safe(tool_name, "unknown", 120)
    parent: Context = extract(_lower(headers))
    tracer = trace.get_tracer("spaceagent.sandbox", "0.1.0")
    with tracer.start_as_current_span(
        f"execute_tool {tool}",
        context=parent,
        kind=SpanKind.INTERNAL,
        attributes={
            "gen_ai.operation.name": "execute_tool",
            "gen_ai.tool.name": tool,
            "gen_ai.tool.type": "function",
            "spaceagent.sandbox.engine": "docker",
            "spaceagent.gen_ai.semconv.commit": GEN_AI_SEMCONV_COMMIT,
        },
    ) as span:
        try:
            yield span
            span.set_status(Status(StatusCode.OK))
        except Exception as error:
            span.set_attribute("error.type", _safe(type(error).__name__, "_OTHER", 120))
            span.set_status(Status(StatusCode.ERROR))
            raise


def _lower(headers: Mapping[str, str]) -> dict[str, str]:
    return {str(key).lower(): str(value) for key, value in headers.items()}


def _safe(value: str | None, fallback: str, maximum: int) -> str:
    candidate = (value or "").strip().lower()
    if not candidate or len(candidate) > maximum:
        return fallback
    if not all(character.isalnum() or character in "._:/-" for character in candidate):
        return fallback
    return candidate
